orbit/pkg/table/mdm/mdm_windows.go

```diff
@@ -0,0 +1,402 @@
+	} else {
+		uint16ptrarr := (*[maxBufSize]uint16)(ptr)[:]
+		return windows.UTF16ToString(uint16ptrarr), nil
+	}
+}
+
+// getEnrollmentInfo returns the MDM enrollment status by calling into OS API IsDeviceRegisteredWithManagement()
+func getEnrollmentInfo() (uint32, string, error) {
+	// variable to hold the MDM enrollment status
+	var isDeviceRegisteredWithMDM uint32 = 0
+
+	// heap-allocated buffer to hold the URI data
+	buffUriData := make([]uint16, 0, maxBufSize)
+	if buffUriData == nil {
+		return 0, "", errors.New("failed to allocate memory for URI data")
+	}
+
+	// IsDeviceRegisteredWithManagement is going to return the MDM enrollment status
+	// https://learn.microsoft.com/en-us/windows/win32/api/mdmregistration/nf-mdmregistration-isdeviceregisteredwithmanagement
+	if returnCode, _, err := procIsDeviceRegisteredWithManagement.Call(uintptr(unsafe.Pointer(&isDeviceRegisteredWithMDM)), maxBufSize, uintptr(unsafe.Pointer(&buffUriData))); returnCode != uintptr(windows.ERROR_SUCCESS) {
+		return 0, "", fmt.Errorf("there was an error calling IsDeviceRegisteredWithManagement(): %s (0x%X)", err, returnCode)
+	} else {
+
+		uriData, err := localUTF16toString(unsafe.Pointer(&buffUriData))
+		if err != nil {
+			return 0, "", err
+		}
+
+		return isDeviceRegisteredWithMDM, uriData, nil
+	}
+}
+
+// executeMDMcommand executes syncML MDM commands against the OS MDM stack and returns the status of the command execution
+func executeMDMcommand(inputCMD string) (string, error) {
+	// Synchronizing MDM command execution
+	mu.Lock()
+	defer mu.Unlock()
+
+	// Enabling MDM command executions
+	if err := enableCmdExecution(); err != nil {
+		return "", err
+	}
+
+	// Close MDM management mutex if neeeded - this is a hack to enable multiple MDM management calls
+	handle, err := windows.OpenMutex(windows.MUTEX_ALL_ACCESS, false, windows.StringToUTF16Ptr(mdmMutexName))
+	if err == nil {
+		windows.CloseHandle(handle) // closing handle just opened due to OpenMutex()
+
+		// then closing previously used handle
+		if err := closeManagementMutex(); err != nil {
+			return "", err
+		}
+	}
+
+	// converting input MDM cmd go string into UTF16 windows string
+	inputCmdPtr, err := windows.UTF16FromString(inputCMD)
+	if err != nil {
+		return "", err
+	}
+
+	// MDM stack is ready to receive commands
+	var outputStrBuffer *uint16
+	if returnCode, _, err := procSendSyncMLcommand.Call(uintptr(unsafe.Pointer(&inputCmdPtr[0])), uintptr(unsafe.Pointer(&outputStrBuffer))); returnCode != uintptr(windows.ERROR_SUCCESS) {
+		return "", fmt.Errorf("there was an error calling ApplyLocalManagementSyncML(): %s (0x%X)", err, returnCode)
+	}
+
+	// converting windows MDM UTF16 output string into go string
+	outputCmd, err := localUTF16toString(unsafe.Pointer(outputStrBuffer))
+	if err != nil {
+		return "", err
+	}
+
+	// freeing OS allocated heap memory
+	_, err = windows.LocalFree(windows.Handle(unsafe.Pointer(outputStrBuffer)))
```

Comment:
Ah I see, so it looks like the `procSendSyncMLcommand.Call` above causes windows (or the windows package anyway) to allocate in the `outputStrBuffer` pointer, so it doesn't have to be pre-allocated.

Reply:
yes, it is heap-data allocated by the OS function, which we need to release to avoid a memleak